### PR TITLE
mystery box SC

### PR DESCRIPTION
**Pull request by psorinionut**

### CODE REVIEW

**Review by alyn509 — APPROVED** [review]

File: contracts/mystery-box/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    #[storage_mapper("mysteryBoxTokenIdentifier")]
+    fn mystery_box_token_identifier(&self) -> SingleValueMapper<TokenIdentifier>;
+
+    #[view(getCooldownEpoch)]
+    #[storage_mapper("cooldownEpoch")]
+    fn cooldown_epoch(&self) -> SingleValueMapper<u64>;
+
+    // #[view(getEgldReserve)]
+    // #[storage_mapper("egldReserve")]
+    // fn egld_reserve(&self) -> SingleValueMapper<BigUint>;
```
Comment (alyn509) [inline contracts/mystery-box/src/config.rs:55]: remove this

**Review by alyn509 — CHANGES_REQUESTED** [review]

File: contracts/mystery-box/src/config.rs (hunk 1 above)
Comment (psorinionut) [reply contracts/mystery-box/src/config.rs]: Removed unused code.

File: contracts/mystery-box/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        let mut accumulated_percentage = 0u64;
+
+        for mystery_box_reward in mystery_box_setup.into_iter() {
+            accumulated_percentage += mystery_box_reward.percentage_chance;
+            self.check_reward_validity(&mystery_box_reward);
+        }
+        require!(
+            accumulated_percentage == MAX_PERCENTAGE,
+            "The total percentage must be 100%"
+        );
```
Comment (sasurobert) [inline contracts/mystery-box/src/lib.rs:53]: Line 46-53 is not needed here. They are needed only when the reward setup is created for the overall SC

File: contracts/mystery-box/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        let mystery_box_token_nonce = self.send().esdt_nft_create_compact(
+            &mystery_box_token_id,
+            &BigUint::from(NFT_AMOUNT),
+            &mystery_box_setup,
+        );
+
+        let user_payment = EsdtTokenPayment::new(
```
Comment (sasurobert) [inline contracts/mystery-box/src/lib.rs:61]: There is no need to create new NFT every time. SFTs are ok, same information, no need to duplicate in separate NFTs.

File: contracts/mystery-box/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            )
+            .decode_attributes();
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let cooldown_epoch = self.cooldown_epoch().get();
+        let mut active_cooldown = true;
+        let mut winning_reward = Reward::default();
+        while active_cooldown {
+            winning_reward = self.get_winning_reward(&mystery_box_attributes);
+            active_cooldown =
```
Comment (sasurobert) [inline contracts/mystery-box/src/lib.rs:102]: What do you want with this while code? 
The SFT of misteryBox should contain an creation epoch. And you should check only that currentepoch - creationEpoch > cooldownPeriod

File: contracts/mystery-box/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                payment.token_nonce,
+            )
+            .decode_attributes();
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let cooldown_epoch = self.cooldown_epoch().get();
+        let mut active_cooldown = true;
+        let mut winning_reward = Reward::default();
+        while active_cooldown {
+            winning_reward = self.get_winning_reward(&mystery_box_attributes);
```
Comment (sasurobert) [inline contracts/mystery-box/src/lib.rs:101]: You have to call this only once.

File: contracts/mystery-box/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            );
+            self.send().direct(
+                &caller,
+                &winning_reward.reward_token_id,
+                winning_reward.reward_token_nonce,
+                &winning_reward.value,
+            );
+        }
+
+        self.send().esdt_local_burn(
```
Comment (sasurobert) [inline contracts/mystery-box/src/lib.rs:125]: I do not see the code for creating a new SFT and sending that to the user

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by alyn509 — APPROVED** [review]

**Review by alyn509 — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by alyn509 — APPROVED** [review]

**Pull request merged**